Create gatsby-image-tutorial (#15027)
gatsbyjs/gatsby: docs/tutorial/gatsby-image-tutorial/gatsby-image-tutorial.md

> _Note: this tutorial uses examples of static content stored in YAML files, but similar methods can be used for Markdown files._

## Getting started

Image optimization in Gatsby is provided by a plugin called `gatsby-image` which is incredibly performant.

### Step 1

Start by using npm to install the `gatsby-image` plugin and its associated dependencies.

```bash
npm install gatsby-image gatsby-transformer-sharp gatsby-plugin-sharp
```

### Step 2

Add the newly installed plugins to your `gatsby-config.js` file. The config file ends up looking like this (other plugins already in use have been removed from this snippet for simplicity).

> _Note: once `gatsby-image` has been installed, it does not need to be included in the `gatsby-config.js` file._

```javascript:title=gatsby-config.js
plugins: [`gatsby-transformer-sharp`, `gatsby-plugin-sharp`]
```

## Gatsby-image configuration

Now you're set up to use `gatsby-image`.

### Step 3

Determine where your image files are located. In this example they're in `src/data`.

If you haven't already, make sure that your project is set up to see content inside that directory. That means doing two things:

> 1.  Install `gatsby-source-filesystem`. Note: If you created your project using `gatsby new <name>`, this first step is already done for you.

Reviewer: ```suggestion
1.  Install `gatsby-source-filesystem`. Note: If you created your project using `gatsby new <name>`, this first step should already be done for you via the default starter.
```

```jsx:title=src/pages/index.js
export const query = graphql`
  query {
    file(relativePath: { eq: "headers/headshot.jpg" }) {
      childImageSharp {
        fixed(width: 125, height: 125) {
          ...GatsbyImageSharpFixed
        }
      }
    }
  }
`
```

There are a couple of things to note here.

### Relative image paths and `gatsby-config.js`

You might expect the relative path to be relative to the file the code sits in, in this case that's index.js. However, that doesn't work. The relative path is actually based on the line of code you put in the `gatsby-source-filesystem` config, which points to `src/data`.

### Image fragments

Another thing to note about this query is how it uses the fragment `GatsbyImageSharpFixed` to return a fixed width and height image. You could also use the fragment `GatsbyImageSharpFluid` which produces scalable images that fill their container instead of fitting specific dimensions. In `gatsby-image`, _fluid_ images are meant for images that don’t have a finite size depending on the screen, where as other images are _fixed_.

> The query will return a processed image at the highest depth. That is, the deepest part of the return object. The returned result will be automatically passed into the component and attached to the `data` prop. You can then display the image using JSX to automatically output responsive, highly performant HTML.

Reviewer: ```suggestion
The query will return a data object including the processed image in a format usable by the `gatsby-image` component. The returned result will be automatically passed into the component and attached to the `data` prop. You can then display the image using JSX to automatically output responsive, highly performant HTML.
```

I think this part will confuse folks–so I'd recommend rewording it slightly. An alternative suggestion would be to show more of the data object returned so the depth part is less abstract.

To display the image, start by importing the component provided by `gatsby-image`.

```jsx
import Img from "gatsby-image"
```

Now you can use it. Note that the key for pointing to the image corresponds to the way in which the image was processed. In this example that is `fixed`.

```jsx
<Img
  className="headshot"
  fixed={data.file.childImageSharp.fixed}
  alt="headshot"
/>
```

Here is the query and usage all put together:

```jsx:title=src/pages/index.js
import React from "react"
import { graphql } from "gatsby"
import Img from "gatsby-image"
import Layout from "../components/layout"

const HomePage = ({ data }) => {
  return (
    <Layout>
      <Img
        className="headshot"
        fixed={data.file.childImageSharp.fixed}

>         alt="headshot"

Reviewer: ```suggestion
        alt=""
```
"headshot" wouldn't be super useful for users of assistive technology, so I'd recommend leaving it blank or pulling in the author/site owner's name.

      />
    </Layout>
  )
}

export const query = graphql`
  query {
    file(relativePath: { eq: "headers/headshot.jpg" }) {
      childImageSharp {
        fixed(width: 125, height: 125) {
          ...GatsbyImageSharpFixed
        }
      }
    }
  }
`
export default HomePage
```

## Querying for multiple images from YAML data

> Another way to source images is through YAML (or Markdown). This example uses the `gatsby-transformer-yaml` plugin to query the YAML files. More information about that plugin can be found in the [Gatsby plugin library](https://www.gatsbyjs.org/packages/gatsby-transformer-yaml/?=gatsby-transformer-yaml).

Reviewer: ```suggestion
Another way to source images is through YAML (or Markdown). This example uses the `gatsby-transformer-yaml` plugin to query the YAML files. More information about that plugin can be found in the [Gatsby plugin library](/packages/gatsby-transformer-yaml/?=gatsby-transformer-yaml).
```

> Here's an example of a query from a list of conferences in a YAML file:

Reviewer: ```suggestion
Here's an example of a query from a list of conferences in a YAML file with an image for each one:
```
In this section, I think it would help to get to the image part sooner so the tutorial doesn't get side-tracked with YAML query information. It would help mitigate the "forking" question you had earlier, @lstar19.

```graphql
{
  allSpeakingYaml {
    edges {
      node {
        conference
        year

>       }

Reviewer: ```suggestion
        image {
          childImageSharp {
            fluid {
              ...GatsbyImageSharpFluid
            }
          }
        }
      }
```
Because the tutorial has already talked about queries, I think it would help to surface the image part of a YAML query sooner

    }
  }
}
```

In this case the query starts with `allSpeakingYaml` to direct `graphql` to look for this data in the `speaking.yaml` file in your `src/data` folder referenced in `gatsby-config.js`. If you want to query a file named `blog.yaml`, for example, you'd start the query with `allBlogYaml`.

### Troubleshooting queries

> A common error you might run into is `Unknown field allSpeakingYaml on type Query`, where `allSpeakingYaml` is replaced with something specific to your code. What the error boils down to is that `graphql` is unable to find a matching result. This can mean it couldn't find a file that matched the first line, meaning that no `speaking.yaml` file exists.

Reviewer: If we can cover the query gotchas in a few bullet points and keep the info limited to images, it will make this more succinct. See comments above about moving the image query stuff up to cover it sooner!

```jsx
{data.talks.edges.map(({ node }) => ())
```

The top-level object name of `data` is implicit. This is important because when you conduct multiple queries as part of a single component, Gatsby still passes the entire result to the component.

Here's an example of data flowing into a component:

```jsx
const SpeakingPage = ({ data}) => {})
```

Everything else gets referenced from that top-level return name.

With that understanding, you can combine two queries referencing images and use aliasing to distinguish between them.

```graphql
{
  allSpeakingYaml {
    edges {
      node {
        image {
          childImageSharp {
            fluid {
              ...GatsbyImageSharpFluid
            }
          }
        }
      }
    }
  }
  banner: file(relativePath: { eq: "headers/default.jpg" }) {
    childImageSharp {
      fluid {
        ...GatsbyImageSharpFluid
      }
    }
  }
}
```

Notice that this example uses aliasing for one query and not the other. This is allowed; there is no requirement that all your queries use aliasing. In this case, the JSX would look like this to access the `speaking.yaml` content.

```jsx

> {data.allSpeakingYaml.edges.map(({ node }) => ())

Reviewer: It would help to surface the image part of this component to keep it relevant to `gatsby-image`.
